components/zcash_address/src/kind/unified.rs

```diff
@@ -208,37 +211,64 @@ impl TryFrom<(&str, &[u8])> for Address {
     type Error = ParseError;
 
     fn try_from((hrp, buf): (&str, &[u8])) -> Result<Self, Self::Error> {
-        let encoded = f4jumble::f4jumble_inv(buf).ok_or(ParseError::InvalidEncoding)?;
+        fn read_receiver(mut cursor: &mut std::io::Cursor<&[u8]>) -> Result<Receiver, ParseError> {
+            let typecode = CompactSize::read(&mut cursor).map_err(|e| {
+                ParseError::InvalidEncoding(format!(
+                    "Failed to deserialize CompactSize-encoded typecode {}",
+                    e
+                ))
+            })?;
+            let length = CompactSize::read(&mut cursor).map_err(|e| {
+                ParseError::InvalidEncoding(format!(
+                    "Failed to deserialize CompactSize-encoded length {}",
+                    e
+                ))
+            })?;
+            let addr_end = cursor.position().checked_add(length).ok_or_else(|| {
+                ParseError::InvalidEncoding(format!(
+                    "Length value {} caused an overflow error",
+                    length
+                ))
+            })?;
+            let addr = cursor.get_ref();
+            if (addr.len() as u64) < length {
```

Comment:
We need to check we can read `length` bytes from `cursor.position()`, not from the beginning (`Cursor::get_ref` returns a reference to the entire underlying data, not the remaining slice):
```suggestion
            if (addr.len() as u64) < addr_end {
```

Reply:
Ahh, that shouldn't be called `addr` then. Fixed.